lancero_source.go

```diff
@@ -468,13 +475,29 @@ func (ls *LanceroSource) distributeData() {
 			lastSampleTime = timeFix
 		}
 	}
-	if framesUsed <= 0 {
-		log.Printf("Nothing to consume, buffer[0] size: %d samples\n", len(buffers[0]))
-		return
-	}
 	timediff := lastSampleTime.Sub(ls.lastread)
 	ls.lastread = lastSampleTime
-
+	// check for changes in nrow, ncol and lsync
+	for ibuf, dev := range ls.active {
+		buffer := buffers[ibuf]
+		q, p, n, err := lancero.FindFrameBits(rawTypeToBytes(buffer))
+		if err != nil {
+			return fmt.Errorf("Error in findFrameBits: %v", err)
+		}
+		ncols := n
+		nrows := (p - q) / n
+		periodNS := timediff.Nanoseconds() / int64(framesUsed)
+		lsync := roundint((float64(periodNS) / 1000) * float64(dev.clockMhz) / float64(nrows))
+		if ncols != dev.ncols || nrows != dev.nrows || lsync != dev.lsync || framesUsed <= 0 {
+			fmt.Printf("have ncols %v, nrows %v, lsync %v, framesUsed %v. had ncols %v, nrows %v, lsync %v, blockingReadCount %v\n",
+				ncols, nrows, lsync, framesUsed, dev.ncols, dev.nrows, dev.lsync, ls.blockingReadCount)
+			if ls.blockingReadCount > 1 {
+				// lsync calculation fails on first few reads so don't error
+				return fmt.Errorf("have ncols %v, nrows %v, lsync %v, framesUsed %v. had ncols %v, nrows %v, lsync %v",
+					ncols, nrows, lsync, framesUsed, dev.ncols, dev.nrows, dev.lsync)
+			}
+		}
+	}
```

Comment:
This block looks expensive. Is it? Did you profile? Also, it looks like it is repeated from somewhere.

Reply:
Also, is it really a good idea, given that LSYNC is computed from timing between reads, which is susceptible to errors when latency is large?

Reply:
I have run a lancero source with this block in place and it worked at 200 MB/s. I haven't profiled it, but I don't think it is expensive unless it is going to fail anyway. `FindFrameBits` only scans through two frames if it is succesful. I am a bit worried we will occsionally get lsync wrong due to random timing variation, and in that case I plan to add a `succesiveErrorsCount` or similar.